### DISCUSSION TITLE
Add full report data structure for templates

### DISCUSSION THREAD

**Discussion by carsongee** [aquasecurity/trivy#4208] [Ideas]
I'd like to use additional `types.Report` data in my templates, currently available only in the JSON report. In particular, `ArtifactName`. It seems that `template.go` only renders from `Results`:

https://github.com/aquasecurity/trivy/blob/79a1ba32d5bec7b868dd80ab20bf7e4c4e8eb5a1/pkg/report/template.go#L75

while JSON takes the full data structure:

https://github.com/aquasecurity/trivy/blob/79a1ba32d5bec7b868dd80ab20bf7e4c4e8eb5a1/pkg/report/json.go#L20

Can we change the `template.go` report to include the full `types.Report` data? I'd happily do the work and refactor the existing `contrib` templates. I know this would have some backward compatibility issues, so I understand if the project prefers to keep it as is. Thanks!

**Comment by knqyf263** [comment]
Thanks for your suggestion. Yes, it brings breaking change and will be a big impact. We were not sure how many users needed metadata in templates, and wondering if it was worth causing confusion. Can we hold it for a while and see the demand from the community?

**Comment by carsongee** [comment]
Yep! I understand, and I'm comfortable with that approach. Our primary use case is collating multiple scans of artifact bundles when scanning Helm charts for example. We can continue to do that with external tooling.

Additionally, the current HTML template uses an awkward 0-index into Targets to title the report that would likely make more sense as the Artifact Name: https://github.com/aquasecurity/trivy/blob/914c6f0921c2811c9a62aad9ae2b2dbd1ae21a0d/contrib/html.tpl#L85.  That can be fixed moving the Target into the results and having a generic report title though. 

As I see it, nothing critical at this point requires it, but I see it getting harder and harder to make this type of change as the project matures if it is ever desired. Regardless, thanks for your consideration and the great project!

**Comment by nicholasdille** [comment]
I am hitting a similar bump here when using templating to create a report for SonarQube.

The existing template (https://github.com/mendhak/trivy-template-output-to-sonarqube) uses `.Target` to populate `.issues[].primaryLocation.filePath`. When the SonarQube scanner attempt to import the generated report, it fails because the file does not exists.

I patched the generated report to double-check this. Without the suffix, SonarQube accepts the report because the filename matches an existing file. The contents of `.Target` seem to come from [pkg/scanner/ospkg/scan.go#L69](https://github.com/aquasecurity/trivy/blob/45d5edb0d70bd7e10b9118038c0cadde9daf2638/pkg/scanner/ospkg/scan.go#L69).

My suggestion would be to add a new field in [pkg/types/report.go#L71](https://github.com/aquasecurity/trivy/blob/45d5edb0d70bd7e10b9118038c0cadde9daf2638/pkg/types/report.go#L71) holding the name of the artifact.

**Reply by fatihtokus** [reply]
Hi @carsongee , @nicholasdille 

I have a trivy plugin(https://github.com/fatihtokus/scan2html) that creates a well-formated filterable/sortable html report for any scan result of trivy. The usage is very simple, please have a look:

Install the plugin: $ trivy plugin install github.com/fatihtokus/scan2html
$ trivy scan2html k8s --report=all cluster result.html

Regards!

**Comment by FideliusFalcon** [comment]
I've ended up here trying to include "artifactName" in my template, but it seems that this field is located in "types.Report" - Is there a fix for this?

**Comment by sjames-au** [comment]
Found my way here too. I was exploring trivy to replace some reports I archive for audits. I generate a report which they prefer reviewing, providing them metadata about the report is helpful making their experience auditing us positive. I'm basically using another language so I can have that meta data in the reports.

**Comment by itaysk** [comment]
while I think the original request is legitimate, I just want to add that there's an alternative solution using [output plugin](https://trivy.dev/latest/docs/plugin/user-guide/#output-mode-support)
